### DISCUSSION TITLE
How to customize or hide the run buttons in editor

### DISCUSSION THREAD

**Discussion by harry-xm** [redhat-developer/lsp4ij#479] [Q&A] [answered]
I'd like to either

1. add a Run with Coverage option and integrate with https://www.jetbrains.com/help/idea/code-coverage.html. Optionally, also put the buttons on the sidebar and use the standard IntelliJ icons.
2. hide the toolbar, so that I can fully customize it in a Rust-specific LSP plugin.

![Screenshot](https://github.com/user-attachments/assets/ce657723-f525-4c73-9fa0-a3a148d30577)

**Comment by angelozerr** [comment, marked as answer]
> add a Run with Coverage option and integrate with https://www.jetbrains.com/help/idea/code-coverage.html. 

To do that, you need to create an IJ plugin and define the LSP command, see https://github.com/redhat-developer/lsp4ij/blob/main/docs/DeveloperGuide.md#lsp-commands

> Optionally, also put the buttons on the sidebar and use the standard IntelliJ icons. 

Indeed for Run and Debug CodeLens it makes totally sense. Badly LSP specification doesn't support that. To support that:

 * LSP specification should be improved. I created the issue at https://github.com/microsoft/language-server-protocol/issues/2009
 * LSP4IJ should provide an API to customize with an IJ plugin the render of CodeLens on the sidebar. Please create an issue for that.

> hide the toolbar, so that I can fully customize it in a Rust-specific LSP plugin.

I think any language server should never report CodeLens mapped with a command if LSP client doesn't support the command. You could fill an issue to rust-analyzer project.

An another idea is to provide an API to hide the Codelens with an IJ plugin. Please create an issue for that.

**Reply by harry-xm** [reply]
> > add a Run with Coverage option and integrate with https://www.jetbrains.com/help/idea/code-coverage.html.
> 
> To do that, you need to create an IJ plugin and define the LSP command, see https://github.com/redhat-developer/lsp4ij/blob/main/docs/DeveloperGuide.md#lsp-commands

How should I add it to the CodeLens then? The language server does not support this feature, so there is no corresponding LSP command. It has to be implemented on client-side only.

**Reply by angelozerr** [reply]
Ok we need to improve documentation to explain that more. Please create an issue.

If you see LSP console, you can see Rust CodeLens:

```
[Trace - 08:33:03] Received response 'textDocument/codeLens - (11)' in 9ms.
Result: [
  {
    "range": {
      "start": {
        "line": 4,
        "character": 3
      },
      "end": {
        "line": 4,
        "character": 7
      }
    },
    "command": {
      "title": "▶︎ Run ",
      "command": "rust-analyzer.runSingle",
      "arguments": [
        {
          "label": "run foo",
          "location": {
            "targetUri": "file:///c:/Users/azerr/test-rust/foo/src/main.rs",
            "targetRange": {
              "start": {
                "line": 4,
                "character": 0
              },
              "end": {
                "line": 10,
                "character": 1
              }
            },
            "targetSelectionRange": {
              "start": {
                "line": 4,
                "character": 3
              },
              "end": {
                "line": 4,
                "character": 7
              }
            }
          },
          "kind": "cargo",
          "args": {
            "workspaceRoot": "C:\\Users\\azerr\\test-rust\\foo",
            "cargoArgs": [
              "run",
              "--package",
              "foo",
              "--bin",
              "foo"
            ],
            "cargoExtraArgs": [],
            "executableArgs": []
          }
        }
      ]
    }
  },
  {
    "range": {
      "start": {
        "line": 4,
        "character": 3
      },
      "end": {
        "line": 4,
        "character": 7
      }
    },
    "command": {
      "title": "Debug",
      "command": "rust-analyzer.debugSingle",
      "arguments": [
        {
          "label": "run foo",
          "location": {
            "targetUri": "file:///c:/Users/azerr/test-rust/foo/src/main.rs",
            "targetRange": {
              "start": {
                "line": 4,
                "character": 0
              },
              "end": {
                "line": 10,
                "character": 1
              }
            },
            "targetSelectionRange": {
              "start": {
                "line": 4,
                "character": 3
              },
              "end": {
                "line": 4,
                "character": 7
              }
            }
          },
          "kind": "cargo",
          "args": {
            "workspaceRoot": "C:\\Users\\azerr\\test-rust\\foo",
            "cargoArgs": [
              "run",
              "--package",
              "foo",
              "--bin",
              "foo"
            ],
            "cargoExtraArgs": [],
            "executableArgs": []
          }
        }
      ]
    }
  }
]

```

You need to define an action with `rust-analyzer.runSingle` id attribute and define your LSPCommandAction class.

Here a sample of LSPCOmmandAction implementation that we use for IJ Quarkus https://github.com/redhat-developer/intellij-quarkus/blob/95206275a64dbe314b51dd8b7c0fd96a1514a35d/src/main/java/com/redhat/devtools/intellij/lsp4mp4ij/psi/core/command/MicroprofileOpenURIAction.java#L21 

In this case Args is filled with a simple String with is a document Uri which opens the file in the editor.

**Reply by harry-xm** [reply]
Perhaps a screenshot would make it more obvious:

![Screenshot PM](https://github.com/user-attachments/assets/03161d13-7ad5-4ab3-a293-a389dcb21ae2)

I want to add a Run with Coverage option, not replacing the Run option. There are `rust-analyzer.runSingle` and `rust-analyzer.debugSingle`, but there isn't a `rust-analyzer.runSingleWithCoverage`.

**Reply by angelozerr** [reply]
Gutter action is not supported by LSP4IJ (you have created the issue at https://github.com/redhat-developer/lsp4ij/issues/483)

So today the only solution is to manage the Run by clicking to the CodeLens:

![image](https://github.com/user-attachments/assets/fb763e11-eb90-44ea-b2fc-df8d82a13170)

Today if you click on this CodeLens, you should have the notification error:

![image](https://github.com/user-attachments/assets/de56c0f6-0e49-40e0-98f8-a3011ba215d0)

because you have not defined the LSPCommandAction with the proper id.

> I want to add a Run with Coverage option, not replacing the Run option. There are rust-analyzer.runSingle and rust-analyzer.debugSingle, but there isn't a rust-analyzer.runSingleWithCoverage

If this codelens doesn't appear it is because the rust-analyzer doesn't support it (or perhaps there is a settings for that). I suggest that you create an issue to rust-analyzer.

**Reply by harry-xm** [reply]
> If this codelens doesn't appear it is because the rust-analyzer doesn't support it (or perhaps there is a settings for that). I suggest that you create an issue to rust-analyzer.

There is https://github.com/rust-lang/rust-analyzer/issues/8881, which should cover this case in some way.

In the meantime, I think I will go with hiding this type of CodeLens and try to use client-side (i.e. IntelliJ platform) only implementation, which depends on https://github.com/redhat-developer/lsp4ij/issues/484.

**Comment by angelozerr** [comment]
> There is https://github.com/rust-lang/rust-analyzer/issues/8881, which should cover this case in some way.

Indeed you will need to wait for a release of rust-analyzer.

> In the meantime, I think I will go with hiding this type of CodeLens and try to use client-side (i.e. IntelliJ platform) only implementation, which depends on https://github.com/redhat-developer/lsp4ij/issues/484.

Why you don't want to implement action?

**Reply by harry-xm** [reply]
Because I don't know how long it will take until a volunteer decides to implement the `E-hard` task, and there is an existing implementation from https://github.com/intellij-rust/intellij-rust that could probably be reused.

This would bring up a separate issue, that is whether IntelliJ's native syntax parsing (a properly parsed PSI tree) and LSP4IJ's fake `LSPPsiElement` could co-exist. Does LSP4IJ use PSI elements in the file at all?

**Reply by angelozerr** [reply]
> Because I don't know how long it will take until a volunteer decides to implement the E-hard task, and there is an existing implementation from https://github.com/intellij-rust/intellij-rust that could probably be reused.

If you want to hide those CodeLens I suggest that you create an issue to rust-analyzer to have a settings to hide Run / Debug. When you will start rust-analyzer you could set this settings and that's it.

> This would bring up a separate issue, that is whether IntelliJ's native syntax parsing (a properly parsed PSI tree) and LSP4IJ's fake LSPPsiElement could co-exist.

It will work, for instance in Intellij Quarkus we haveQute PsiFile (to support syntax coloration) and Qute LS (for completion, validation, etc) and it works great. The only problem that I see today is if the Rust PsiFile doesn't tokenize correctly the Rust code and you could have some trouble with completion like https://github.com/redhat-developer/lsp4ij/issues/447

Implementing LSP command is very easy and having Run / Debug commands is very nice too.

**Reply by harry-xm** [reply]
Another feature that IntelliJ native implementations have is Run/Debug actions in editor's context menu (e.g. right click in a main function's body), which might be difficult to implement in LSP's CodeLens model.

![Screenshot](https://github.com/user-attachments/assets/dded3898-2db5-4fe0-9937-f79cce859941)

**Comment by angelozerr** [comment]
@harry-xm please try the LSP API PR https://github.com/redhat-developer/lsp4ij/pull/543 which should give the capability to ignore some CodeLens for a given language server.
